### PR TITLE
Instructor: moderate feedback: give message "Some questions may be hidden..." as a status message #2818

### LINKED ISSUE #2818: Instructor: moderate feedback: give message "Some questions may be hidden..." as a status message

### LINKED ISSUE #5932: Moderating responses/previewing sessions heading does not properly escape HTML characters
When moderating a student's responses, the heading of the page does not properly escape HTML, resulting in HTML elements possibly being introduced into the heading.
sani

Edit: This also applies to previewing sessions.

### PR DESCRIPTION
**Pull request by xpdavid**
Fixes #2818 
Fixes #5932

**Changes in code:**
show question visibility message in the top bar: 16285c21c6621be1e30188f121a45d1005cd7b88

**Test cases fixed:**
InstructorEditStudentFeedbackPageUiTest: 0e034e09bb01091f4ac728c7e8a8045bdfd6804f

Before:
screen shot 2017-01-14 at 11 07 31 pm

After:
screen shot 2017-01-15 at 2 45 42 pm

Needs help in phrasing of the second sentence (discussion [here](https://github.com/TEAMMATES/teammates/issues/2818#issuecomment-272676999)). 

@damithc I have made it relatively better.

### CODE REVIEW

**Comment by damithc** [comment]
@JosephineKwa please delay the review until I figure out the details of the issue.

**Comment by xpdavid** [comment]
@damithc @JosephineKwa Ready for review.
May need discussion on the phrasing of the second sentence.

File: src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSubmissionForm.tag [hunk 1]
```diff
@@ -23,14 +23,6 @@
     <t:statusMessage statusMessagesToUser="${data.statusMessagesToUser}" />
     <tsfse:feedbackSessionDetailsPanel feedbackSession="${data.bundle.feedbackSession}"/>
     
-    <c:if test="${data.moderation}">
-        <div class="row">
-            <span class="help-block align-center">
-                <%= Const.FEEDBACK_SESSION_QUESTIONS_HIDDEN %>
```
Comment (JosephineKwa) [inline src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSubmissionForm.tag]: Do remove this constant from `Const.java` if it is not used anywhere else

File: src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSubmissionEdit.tag [hunk 2]
```diff
@@ -23,7 +23,22 @@
                     <h3 class="text-center">Previewing Session as ${isInstructor ? "Instructor" : "Student"} ${moderatedPersonName} (${moderatedPersonEmail})</h3>
                 </c:when>
                 <c:when test="${data.moderation}">
-                    <h3 class="text-center">Moderating Responses for ${isInstructor ? "Instructor" : "Student"} ${moderatedPersonName} (${moderatedPersonEmail})</h3>
+                    <div class="container">
+                        <div class="col-md-12">
+                            <h3 class="text-center">
+                              You are moderating responses for (or submitting on behalf of) ${isInstructor ? "Instructor" : "Student"} ${moderatedPersonName} (${moderatedPersonEmail})
```
Comment (JosephineKwa) [inline src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSubmissionEdit.tag]: For consistency, let's keep the indentation as 4 spaces, i.e.
```
<h3 class="text-center">
    You are ...
```

Comment (xpdavid) [inline src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSubmissionEdit.tag]: @JosephineKwa 
Sure, I'll change. But is it necessary to update the HTML coding standard [documentation](https://oss-generic.github.io/process/codingStandards/CodingStandard-Html.html) in the Miscellaneous section. It actually uses two spaces for content.

**Comment by damithc** [comment]
> It covers the "Submit Feedback" header.

I didn't realize the banner is covering stuff underneath it. That's not good. Yes, should try to avoid it one way or other. 

Actually, moderating is different from 'submitting on behalf' (i.e. submitting on behalf of a student who missed the deadline) but it's ok not to mention it here. When the user is here, he already knows what he is doing, so no need to be elaborate about it.

File: src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSubmissionEdit.tag (hunk 2 above)
Comment (JosephineKwa) [inline src/main/webapp/WEB-INF/tags/shared/feedbackSubmissionEdit/feedbackSubmissionEdit.tag]: Yup the documentation does specify that indentation should be 2 spaces, and indeed we should be following that 😄 

It would be good to update the indentation for the rest of the code to using 2 spaces to avoid having the html skewed towards the right in cases where there are many nested tags. However that's a fairly monumental task. 

As most other parts of the code, for example `index.jsp` use 4 spaces for indentation, to maintain consistency, we stick to that as the amount of spaces to use for now.

**Comment by xpdavid** [comment]
@damithc @JosephineKwa 

I make a "Show more" link, by default the moderation hint is not shown.

screen shot 2017-01-16 at 11 11 50 pm

--- 

screen shot 2017-01-16 at 11 11 55 pm

**Comment by damithc** [comment]
@xpdavid It's common to pair 'show more' with 'show less'. 

Shall we use 'show more details' and 'show less details'?

**Comment by JosephineKwa** [comment]
Hmm I feel that "Show more" and "Show less" should be sufficient in indicating to the user that he can click it to learn more. In fact, some other parts of TEAMMATES, e.g. the results page, currently just uses "more" and "less".

more
less

**Comment by xpdavid** [comment]
**Changes in code:**

- remove unused constant in Java: 58f20c57bac78f3729a15934895cf6d244e01cb7
- add close/show button to moderation hint: c146020f02fb1d907a42685dde60bfdd395ad0c1

**UI demo:**
screen shot 2017-01-17 at 12 04 52 am

---

screen shot 2017-01-17 at 12 04 56 am

**Comment by damithc** [comment]
> Hmm I feel that "Show more" and "Show less" should be sufficient in indicating to the user that he can click it to learn more. In fact, some other parts of TEAMMATES, e.g. the results page, currently just uses "more" and "less".

True. Even 'more' and 'less' is fine.

File: src/main/webapp/js/feedbackSubmissionsEdit.js [hunk 3]
```diff
@@ -113,6 +115,32 @@ function isPreview() {
     return $(document).find('.navbar').text().indexOf('Preview') !== -1;
 }
 
+function isModeration() {
+    return $('#moderationHintButton').length !== 0;
+}
+
+function bindModerationHintButton() {
+    if (!isModeration()) {
```
Comment (JosephineKwa) [inline src/main/webapp/js/feedbackSubmissionsEdit.js]: I was just wondering, could we simply do the following? 🤔 
```
if (!isModeration()) {
    var expandText = '[Show more]';
    ...
```
To me that seems more straightforward, as it tells devs that the method body will be executed only if user is currently moderating

File: src/main/webapp/js/feedbackSubmissionsEdit.js [hunk 4]
```diff
@@ -113,6 +115,32 @@ function isPreview() {
...
+        return;
+    }
+    
+    var expandText = '[Show more]';
+    var closeText = '[Show less]';
+    $('#moderationHintButton').text(expandText);
+    
+    $('#moderationHintButton').click(function(event) {
```
Comment (JosephineKwa) [inline src/main/webapp/js/feedbackSubmissionsEdit.js]: Since we use `$('#moderationHintButton')` multiple times here, it is good to extract it into the variable `$moderationHintButton`

File: src/main/webapp/js/feedbackSubmissionsEdit.js [hunk 5]
```diff
@@ -113,6 +115,32 @@ function isPreview() {
...
+}
+
+function bindModerationHintButton() {
+    if (!isModeration()) {
+        return;
+    }
+    
+    var expandText = '[Show more]';
+    var closeText = '[Show less]';
+    $('#moderationHintButton').text(expandText);
```
Comment (JosephineKwa) [inline src/main/webapp/js/feedbackSubmissionsEdit.js]: Rather than adding the default text by javascript, my preference would be that the default text is written directly in the `<a>` tag in the `feedbackSubmissionEdit.tag` file. This is so that devs know the default text simply by looking at the `.tag` file rather than having to reference the `.js` file.

Comment (xpdavid) [inline src/main/webapp/js/feedbackSubmissionsEdit.js]: @JosephineKwa I am considering that if later the text is going to be changed, one could simply change the text in `js` file. Since the text `<a>` in are going to be changed between `[More]` and `[Less]`, by putting the text inside `js` file could be more concise.

**Comment by JosephineKwa** [comment]
I missed this out earlier, but the PR title should match the Issue title, i.e. "Instructor: moderate feedback: give message "Some questions may be hidden..." as a status message #2818". Do update this as well

File: src/main/webapp/js/feedbackSubmissionsEdit.js (hunk 3 above)
Comment (xpdavid) [inline src/main/webapp/js/feedbackSubmissionsEdit.js]: @JosephineKwa That would result the whole logic wrapping inside the if statement. I am trying to use guard statement to avoid this (to separate logic and error handling in some sense).

**Comment by xpdavid** [comment]
@JosephineKwa Ready for review

**Comment by JosephineKwa** [comment]
I think you missed out the Issue number in the issue title 😮

**Comment by xpdavid** [comment]
@JosephineKwa Test cases added

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java [hunk 6]
```diff
@@ -35,11 +36,25 @@ public static void classSetup() {
...
     }
     
+    private void testModerationHint() throws Exception {
+        ______TS("verify moderation hint");
+        
+        submitPage = loginToInstructorEditStudentFeedbackPage("IESFPTCourseinstr", "student1InIESFPTCourse@gmail.tmt",
+                "session1InIESFPTCourse");
+        
+        submitPage.verifyHtmlPart(By.className("navbar"), "/instructorEditStudentFeedbackHintLess.html");
+        submitPage.click(By.id("moderationHintButton"));
```
Comment (JosephineKwa) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: Would be good to extract this into a method `clickModerationHintButton()` in `FeedbackSubmitPage.java`

Comment (xpdavid) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: `FeedbackSubmitPage.java` also used by `StudentFeedbackSubmitPageUiTest.java`.
I am thinking that since the `moderationHintButton` is only used by the Instructor's Moderation page, it may not be appropriate for `FeedbackSubmitPage.java .java` to hold the method `clickModerationHintButton()`. May put in the `InstructorEditStudentFeedbackPageUiTest` instead

Comment (xpdavid) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: @JosephineKwa So we put it in `FeedbackSubmitPage.java` right?

Comment (xpdavid) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: A more appropriate method would be created another page `InstructorEditStudentFeedbackPage` which inherit from `FeedbackSubmitPage` and put the method there. But this could be another PR.

Comment (JosephineKwa) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: Sorry I "👍"ed  the wrong comment. 

> A more appropriate approach would be creating another page `InstructorEditStudentFeedbackPage` which inherit from `FeedbackSubmitPage` and put the method there. But this could be another PR.

This approach sounds good.

Comment (xpdavid) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: Can I do it in another issue and PR as this issue is related to code quality but has nothing to do with current PR title.

Comment (JosephineKwa) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: Sure! Go ahead :)

**Comment by JosephineKwa** [comment]
@xpdavid is this ready for review?

**Comment by xpdavid** [comment]
@JosephineKwa Yes. I extracted the method already.

File: src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java [hunk 7]
```diff
@@ -196,4 +198,22 @@ public void waitForOtherOptionTextToBeClickable(int qnNumber, int responseNumber
...
         waitForElementToBeClickable(element);
     }
+    
+    // ------------- For InstructorEditStudentFeedbackPage -------------
+    
+    public boolean clickModerationHintButton() {
+        click(By.id("moderationHintButton"));
+        // Check if links toggle properly.
+        WebElement moderationHint = browser.driver.findElement(By.id("moderationHint"));
+        return moderationHint.isDisplayed();
```
Comment (JosephineKwa) [inline src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java]: It doesn't seem like a good idea to check visibility of the moderation hint through the method used to click on the button. Do extract the moderation hint visibility check into a method specifically for that purpose.

Additionally, for the visibility check, there is an existing method in `AppPage.java` that is provided, `isElementVisible()`.

Comment (xpdavid) [inline src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java]: I am trying to follow the style in `InstructorFeedbackResultsPageUiTest`, it test the similar situation where the panel is toggled and the text box changed.
``` java
assertEquals("[more]", resultsPage.getQuestionAdditionalInfoButtonText(8, "section-1-giver-1-recipient-1"));
        assertTrue(resultsPage.clickQuestionAdditionalInfoButton(8, "section-1-giver-1-recipient-1"));
        assertEquals("[less]", resultsPage.getQuestionAdditionalInfoButtonText(8, "section-1-giver-1-recipient-1"));
        assertFalse(resultsPage.clickQuestionAdditionalInfoButton(8, "section-1-giver-1-recipient-1"));
        assertEquals("[more]", resultsPage.getQuestionAdditionalInfoButtonText(8, "section-1-giver-1-recipient-1"));
```
If this is not acceptable, we need another issue and PR to make the test case `InstructorFeedbackResultsPageUiTest` better as well.

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java [hunk 8]
```diff
@@ -35,11 +35,29 @@ public void classSetup() {
...
     }
     
+    private void testModerationHint() throws Exception {
+        ______TS("verify moderation hint");
+        
+        submitPage = loginToInstructorEditStudentFeedbackPage("IESFPTCourseinstr", "student1InIESFPTCourse@gmail.tmt",
+                "session1InIESFPTCourse");
+        
+        submitPage.verifyModerationHeaderHtml("/instructorEditStudentFeedbackHint.html");
+
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorEditStudentFeedbackPageUiTest.java]: The line `assertEquals("[More]", submitPage.getModerationHintButtonText());` is deleted here as `submitPage.verifyModerationHeaderHtml("/instructorEditStudentFeedbackHint.html");` already verified the text in the link.

File: src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java (hunk 7 above)
Comment (JosephineKwa) [inline src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java]: Indeed the test case in `InstructorFeedbackResultsPageUiTest` can be improved. While it is good to reference the existing code to see how things are currently being done, there are some parts which still suffer from code quality issues 😮 . In the above case, the method is used to click on an element, and simply it should just be doing that.

Comment (xpdavid) [inline src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java]: @JosephineKwa Changed Currently version seems more clear. Should I create another issue on `InstructorFeedbackResultsPageUiTest` to refactor the test case as well :)

Comment (JosephineKwa) [inline src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java]: Yup, do go ahead 😄

**Comment by xpdavid** [comment]
@JosephineKwa Now should be better. :)

**Comment by JosephineKwa** [comment]
Great work 👍 Pending approval from @damithc

File: src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java [hunk 9]
```diff
@@ -196,4 +198,23 @@ public void waitForOtherOptionTextToBeClickable(int qnNumber, int responseNumber
...
+    public boolean isModerationHintVisible() {
+        return isElementVisible("moderationHint");
+    }
+    
+    public void verifyModerationHeaderHtml(String filePathParam) throws IOException {
+        verifyHtmlPart(By.className("navbar"), filePathParam);
+    }
+    
+    public String getModerationHintButtonText() {
+        WebElement moderatorionHintButton = browser.driver.findElement(By.id("moderationHintButton"));
```
Comment (JosephineKwa) [inline src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java]: Oops, missed this out:
`moderatorionHintButton` --> `moderationHintButton`

**Comment by xpdavid** [comment]
@JosephineKwa Should be fine now. I am sorry for those stupid errors. :(

**Comment by JosephineKwa** [comment]
@xpdavid it's all good, we all make these errors from time to time 😄
